### PR TITLE
Bug fixes and experiment

### PR DESCRIPTION
**Pull request by BjoerklundJacob**

### CODE REVIEW

**Review by Zzeeker — DISMISSED** [review]
Boom

File: Simulation/Pathfinding/ActionPathAux.cpp [hunk 1]
```diff
@@ -31,10 +33,23 @@ Path ActionPathAux::getPath() const{
     return path;
 }
 
+Path ActionPathAux::getPath(const Action& lastAction) const{
+    if (this->action == lastAction){
+        return this->getPath();
+    }
+    std::shared_ptr<ActionPathAux> predecessor = this->predecessor;
```
Comment (Simson475) [inline Simulation/Pathfinding/ActionPathAux.cpp:45]: skriv om. vi har snakket om det :)

File: Simulation/Pathfinding/Experiments/CMakeLists.txt [hunk 2]
```diff
@@ -42,13 +42,15 @@ set(PathfindingFiles
 )
 
 # Emit all warning messages
-add_definitions(-Wall -O3)# -Werror
+add_definitions(-Wall -O3 -g)# -Werror
```
Comment (Simson475) [inline Simulation/Pathfinding/Experiments/CMakeLists.txt:45]: fjern -g

**Review by Suprezer — APPROVED** [review]

File: Simulation/Pathfinding/HighLevelCBS.hpp [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 class HighLevelCBS;
 
-#include "argos_wrapper/argos_wrapper.hpp"
+//#include "argos_wrapper/argos_wrapper.hpp"
```
Comment (Simson475) [inline Simulation/Pathfinding/HighLevelCBS.hpp]:
```suggestion
```

File: Simulation/Pathfinding/LowLevelCBS.cpp [hunk 4]
```diff
@@ -23,13 +23,24 @@ Path LowLevelCBS::getIndividualPath(std::shared_ptr<Graph> graph, AgentInfo agen
     std::shared_ptr<Vertex> u;
     Action firstAction = agent.getCurrentAction();
     std::shared_ptr<Vertex> goal = agent.getGoal();
+    bool hasWorked = false;
+    Action workAction = Action(-1, u, u, -1);
```
Comment (Simson475) [inline Simulation/Pathfinding/LowLevelCBS.cpp]: fix

File: Simulation/Pathfinding/LowLevelCBS.cpp [hunk 5]
```diff
@@ -60,30 +73,41 @@ Path LowLevelCBS::getIndividualPath(std::shared_ptr<Graph> graph, AgentInfo agen
...
+            Error::log(top.getPath().toString() + "\n");
+            #endif
             return top.getPath();
         }
 
         // Expand frontier
-        std::vector<Action> possibleActions = getPossibleActions(u, constraints, top.action.timestamp + top.action.duration);
+        std::vector<Action> possibleActions = getPossibleActions(u, constraints, top.action.timestamp + top.action.duration, pathEndTime);
         for (Action action : possibleActions){
+            // If this action can be a final action (after working) return the path immediately
```
Comment (Simson475) [inline Simulation/Pathfinding/LowLevelCBS.cpp:99]: yeet

File: Simulation/Pathfinding/LowLevelCBS.cpp [hunk 6]
```diff
@@ -23,13 +23,24 @@ Path LowLevelCBS::getIndividualPath(std::shared_ptr<Graph> graph, AgentInfo agen
     std::shared_ptr<Vertex> u;
     Action firstAction = agent.getCurrentAction();
     std::shared_ptr<Vertex> goal = agent.getGoal();
+    bool hasWorked = false;
+    Action workAction = Action(-1, u, u, -1);
+    float pathEndTime = 0;
```
Comment (Simson475) [inline Simulation/Pathfinding/LowLevelCBS.cpp]: rename

**Review by Simson475 — CHANGES_REQUESTED** [review]

File: Simulation/Pathfinding/LowLevelCBS.cpp [hunk 7]
```diff
@@ -23,13 +23,24 @@ Path LowLevelCBS::getIndividualPath(std::shared_ptr<Graph> graph, AgentInfo agen
...
+    for (auto& c : constraints){
+        pathEndTime = std::max(c.timeEnd, pathEndTime);
```
Comment (Zzeeker) [inline Simulation/Pathfinding/LowLevelCBS.cpp]: LastConstraintEndTime?

File: Simulation/Pathfinding/LowLevelCBS.cpp [hunk 8]
```diff
@@ -151,7 +176,7 @@ std::vector<Action> LowLevelCBS::getPossibleActions(std::shared_ptr<Vertex> vert
         for (Constraint &constraint : constraints){
             // Edge constraints
             if (ConstraintUtils::isViolatingConstraint(constraint, edge, currentTime)){
-                minWaitTime = std::min(minWaitTime, constraint.timeEnd + DELTA);
+                minWaitTime = std::min(minWaitTime, constraint.timeEnd - currentTime + DELTA);
```
Comment (Zzeeker) [inline Simulation/Pathfinding/LowLevelCBS.cpp:175]: Possibly constraint.timeStart, though shouldnt change much

**Review by Simson475 — APPROVED** [review]

**Pull request merged**